### PR TITLE
Extract DSU from Kruskal code

### PR DESCRIPTION
**Pull request by pdnm**
I think we should use 1 implementation for consistency @howsiwei

### CODE REVIEW

**Comment by stevenhalim** [comment]
Agree, I will let @howsiwei (ch2 PIC) do this final check before merging.

File: ch2/08_unionfind_ds.ml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  
+  let num_disjoint_sets { num_sets } = !num_sets
+
+  let size_of_set dsu u = dsu.size.(find dsu u)
+end
+
+
+open Printf
+
+let () =
```
Comment (howsiwei) [inline ch2/08_unionfind_ds.ml:50]: The cpp version contains `printf("Assume that there are 5 disjoint sets initially\n");` so I guess add that too.

File: ch2/08_unionfind_ds.ml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  let open UnionFind in
+  let dsu = create 5 in
+  let print_nsets () = printf "%d\n" (num_disjoint_sets dsu) in
+  print_nsets();
+  [0, 1; 2, 3; 4, 3] |> List.iter (fun (u, v) ->
+    union dsu u v;
+    print_nsets()
+  );
+  [0, 3; 4, 3] |> List.iter (fun (u, v) ->
+    printf "isSameSet(%d, %d) = %b\n" u v (is_same_set dsu u v)
```
Comment (howsiwei) [inline ch2/08_unionfind_ds.ml:60]: The ocaml version prints `false`/`true` whereas the cpp version prints `0`/`1`. Not sure how important does consistency matter. @stevenhalim what's your opinon?

File: ch2/08_unionfind_ds.ml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  let union dsu u v =
+    let link x p =
+      dsu.parent.(x) <- p;
+      dsu.size.(p) <- dsu.size.(p) + dsu.size.(x);
+      if dsu.rank.(x) = dsu.rank.(p) then
+        dsu.rank.(p) <- dsu.rank.(p) + 1
+    in
+    let pu, pv = find dsu u, find dsu v in
+    if pu = pv then ()
+    else begin
```
Comment (howsiwei) [inline ch2/08_unionfind_ds.ml:36]: This is just a minor style issue but maybe change to `if pu <> pv then begin ...`.
Also note that the cpp version use `isSameSet` but this version is probably better.

**Comment by stevenhalim** [comment]
ok let's merge

**Pull request merged**